Design by contract (#33)
SW814F20/Report: report/CommunicationStrategies/usecases.tex

\section{Requirements model}

> As a methodology for the customers ability to create contract that the system under development (SUD) should uphold we will look at a model for specifying requirements.

Reviewer: ```suggestion
As a methodology for the customers ability to create a contract that the system under development (SUD) should uphold, we will look at a model for specifying requirements.
```

> More specificly we will consider the requirements model as presented in the book "Object-Oriented Software Engineering - A Use Case Driven Approach" by Ivar Jacobson [ref to book].

Reviewer: Indsæt ref til bog

> We will not focus on the the other models from Jacobsons recommended development process, but the requirements model is perticular usefull for our solution and therefore we will focus on how he uses the notion of usecases and actors and how that can be adobted into our solution.

Reviewer: ```suggestion
We will not focus on the the other models from Jacobsons recommended development process, but the requirements model is perticular useful for our solution and therefore we will focus on how he uses the notion of use cases and actors and how that can be adopted into our solution.
```

Jacobson specifies two entities in formulating requirements, \textbf{Actors} and \textbf{Usecases}.

> An \textbf{actor} represents a role that a user have in relation to the system, a user might take different roles when interacting with the system, thereby an \textbf{Actor} is not linked directly with a user but is more a category of users.

Reviewer: Ikke nødvendigt med nr 2 textbf, da du har introduceret actor
```suggestion
An \textbf{actor} represents a role that a user have in relation to the system, a user might take different roles when interacting with the system, thereby an Actor is not linked directly with a user, but is more a category of users.
```

> A \textbf{usecase} represent something that the users should be able to do with in the system. Each \textbf{usecase} takes the perspective of the user and is formulated as a complete list of events in the interaction with the system when utilizing the functionality regarded in the usecase.

Reviewer: Same as above
```suggestion
A \textbf{use case} represent something that the users should be able to do with in the system. Each use case takes the perspective of the user and is formulated as a complete list of events in the interaction with the system when utilizing the functionality regarded in the use case.
```

\subsection{Actors}

> In more general terms a actor is something that can interact with the system. That could be a person, a machine or another software running on the same machine. 

Reviewer: ```suggestion
In more general terms an actor is something that can interact with the system. That could be a person, a machine or another software running on the same machine. 
```

> A key destingtion to make is that a, ex. person, is not limited to be only one type of actor.

Reviewer: destingtion er ikke et rigtigt ord. Lidt i tvivl om hvad det skal være :)

> A key destingtion to make is that a, ex. person, is not limited to be only one type of actor.

Reviewer: ```suggestion
A key destingtion to make is that, a person, is not limited to be only one type of actor.
```

Actors should more be regarded as roles in the system, so if Britta works at a store as a clerk, she will have the actor type clerk when she is working, but in her freetime she might shop in the same store and thereby take the actor role as customer. 

In a broader sense we can consider actors to be everything communicating with the system from outside the system it self.

> In our solution we can use the notion of actors to identify roles with in the SUD, this will allow the developers and customers to communicate easier since they will have a unified vocabulary when talking about who communicates with the system. 

Reviewer: ```suggestion
In our solution we can use the notion of actors to identify roles within the SUD- 
This will allow the developers and customers to communicate easier since they will have an unified vocabulary when talking about who communicates with the system. 
```

> It will also help if the SUD needs boundaries with in the system, i.e. functionalities only avaible to some actors, the customers will then be able to specify a list of actors who should have access to said functionalities, which will help the developers in ensuring correctness.

Reviewer: ```suggestion
It will also help if the SUD needs boundaries within the system, i.e. functionalities only available to some actors.
The customers will then be able to specify a list of actors who should have access to said functionalities, which will help the developers in ensuring correctness.
```

> The developers might even create a access layer with in the SUD based of the actors identified in the requirements or groupings of actors.

Reviewer: ```suggestion
The developers might even create an access layer within the SUD based of the actors identified in the requirements or groupings of actors.
```

> \subsection{Usecases}

Reviewer: ```suggestion
\subsection{Use cases}
```

> A usecase is a complete set of events that specifies all the action needed for a perticular actor to achieve a goal with in the system. 

Reviewer: ```suggestion
A use case is a complete set of events that specifies all the action needed for a particular actor to achieve a goal within the system. 
```

> The set of events should be considered as transaktions - meaning, first the actor does something, then the system respons, then the actor does something, then the system respons and so forth. 

Reviewer: ```suggestion
The set of events should be considered as transactions - meaning, first the actor does something, then the system respons, then the actor does something, then the system respons and so forth. 
```

> A usecase is always from the perspective of a perticular actor, therefore it is important to only focus on details relevant for said actor. 

Reviewer: ```suggestion
A use case is always from the perspective of a particular actor, therefore it is important to only focus on details relevant for said actor. 
```

> A actor will have many usecases and a there will be many usecases similiar to one or another - that could be multiple actors having similar usecases or many usecases having similar event traces. 

Reviewer: ```suggestion
An actor will have many use cases and there will be many use cases similiar to one or another - there could be multiple actors having similar use cases or many use cases having similar event traces. 
```

> The act of defining usecases can help discover requirements for the system, since it forces one to think about how each actor is suppose to interact with the system.

Reviewer: ```suggestion
The act of defining use cases can help discover requirements for the system, since it forces one to think about how each actor is suppose to interact with the system.
```

> Since the usecases are a complete trace of events for a perticular action or goal, a single usecase can then be considered as a transaction system describing a feature og functionality of the system. 

Reviewer: ```suggestion
Since the use cases are a complete trace of events for a particular action or goal, a single use case can then be considered as a transaction system describing a feature or functionality of the system. 
```

> The group of all usecases then specifies all the features and functionalities of the system and how they should work. 

Reviewer: ```suggestion
The group of all use cases then specifies all the features and functionalities of the system and how they should work. 
```

This allows us to consider our software as a set of state transition graphs where stimulus between the actors and the system determine the behaviour of the system.

> The notion of usecases does not fit perfectly into our project, since we suspect that it would be impossible for a customer to know how to construct the event traces in a realistic manner. 

Reviewer: ```suggestion
The notion of use cases does not fit perfectly into our project, since we suspect that it would be impossible for a customer to know how to construct the event traces in a realistic manner. 
```

> But we can get a lot of inspiration by the notion of usecases. 

Reviewer: ```suggestion
However, we can get a lot of inspiration by the notion of use cases. 
```

> Consider if the customer could specify that "As a [actor]. When I..., then I should be able to..., such that..", if we were able to facilitate such requirements in our system we already have alot of information, firstly we can reason about which actors should have access to which components.

Reviewer: ```suggestion
Consider if the customer could specify that "As a [actor]. When I..., then I should be able to..., such that..".
If we were able to facilitate such requirements in our system, we already have much information.
Firstly we can reason about which actors should have access to which components, and can reason about which functionalities which components should be exposed to the different actors.
```

> We can reason about which functionalities of which components should be exposed to the different actors.

Reviewer: Har skrevet den sammen med linjen ovenover
```suggestion
```

> Furthermore the format starts to feel a bit like the "Given, when, then"-model discussed in~\autoref{sec:DbC}, a reader familiar with the agile process Scrum might also recognise that it is very similiar to the User storie template often used in Scrum.

Reviewer: ```suggestion
Furthermore, the format starts to feel a bit like the "Given, when, then"-model discussed in~\autoref{sec:DbC}.
A reader familiar with the agile process Scrum might also recognise that it is very similiar to the, User Stories template, often used in Scrum.
```

\section{Requirements model}

As a methodology for the customers ability to create a contract that the system under development (SUD) should uphold, we will look at a model for specifying requirements.
More specificly we will consider the requirements model as presented in the book "Object-Oriented Software Engineering - A Use Case Driven Approach" by Ivar Jacobson~\cite{Jacobson1992}.

We will not focus on the other models from Jacobsons recommended development process, but the requirements model is perticular useful for our solution and therefore we will focus on how he uses the notion of use cases and actors and how that can be adopted into our solution.

> Jacobson specifies two entities in formulating requirements, \textbf{Actors} and \textbf{Usecases}.

Reviewer: ```suggestion
Jacobson specifies two entities in formulating requirements, \textbf{Actors} and \textbf{Use cases}.
```

An \textbf{actor} represents a role that a user has in relation to the system, a user might take different roles when interacting with the system, thereby an Actor is not linked directly with a user, but is more a category of users.
A \textbf{use case} represents something that the users should be able to do within the system. Each use case takes the perspective of the user and is formulated as a complete list of events in the interaction with the system when utilizing the functionality regarded in the use case.

\subsection{Actors}
In more general terms an actor is something that can interact with the system. That could be a person, a machine, or another software running on the same machine. 
A key distinction to make is that a person is not limited to be only one type of actor.
Actors should more be regarded as roles in the system, so if Britta works at a store as a clerk, she will have the actor type clerk when she is working, but in her free time, she might shop in the same store and thereby take the actor role as a customer. 

In a broader sense, we can consider actors to be everything communicating with the system from outside the system itself.

> In our solution, we can use the notion of actors to identify roles within the SUD- 

Reviewer: ```suggestion
In our solution, we can use the notion of actors to identify roles within the SUD.
```
